### ISSUE TITLE
Visualize synopsis of large graphs

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1991]
People appreciate our `dot` images, especially when coupled with visualizations like the following (courtesy of Jim)

![](http://jcrist.github.io/images/grid_search_schedule.gif)

Unfortunately these don't scale beyond a few hundred nodes.  Meaningfully visualizing very large graphs is a *hard problem*.  

However, while we can't visualize full dask graphs, we might be able to build synopses graphs and visualize those.  In particular, we may be able to take advantage of Dask's naming scheme to show connections between groups of tasks.  Presumably line widths and node sizes/shapes could be used to fill in some of the dropped information.

Additionally, if we can flesh out a good way to construct and visualize these on the fly then we may be able to get something live running on Bokeh.  This is probably secondary to just visualizing things statically though.

cc @eriknw , who just gave a cool talk/demo about visualizing graphs

**Comment by JohnCrickett** [comment]
Perhaps for a browser based option borrow some ideas from Google Map's clustering: https://developers.google.com/maps/documentation/javascript/marker-clustering I've had some success in the past using this to visualise very large telco networks allowing a high level view that can be zoomed in on for detail.

**Comment by mrocklin** [comment]
@eriknw I'd like to engage your thinking a bit here if you get a chance.  I'm considering building a function to produce the following data:

```python
def test_synopsis():
    dsk = {'x': (inc, 1),
           'y-1': (add, 'x', 1),
           'y-2': (add, 'x', 2),
           'y-3': (add, 'x', 3),
           'total': (sum, ['y-1', 'y-2', 'y-3']),
           'z-1': (add, 'total', 'y-1'),
           'z-2': (add, 'z-1', 'y-2'),
           'z-3': (add, 'z-2', 'y-3')}

    nodes, edge_counts = summarize(dsk)
    assert edge_counts == {'y': ('x', 3),
                           'total': ('y', 3),
                           'z': ('total', 1),
                           'z': ('y', 3)}

    assert nodes == {'x': {'count': 1, 'in': 0, 'out': 3},
                     'y': {'count': 3, 'in': 1, 'out': 4},
                     'total': {'count': 1, 'in': 3, 'out': 1}, 
                     'z': {'count': 3, 'in': 4, 'out': 0, 'self': 2}} 
```

Given your experience with graph visuals and your experience with metrics on dask graphs do you have any thoughts on 

1.  What data we should collect
2.  How we should display that data

**Comment by eriknw** [comment]
Ok.  Thinking.

**Comment by shoyer** [comment]
TensorFlow actually solves this problem pretty well with it's graph visualization tools.

The key element that makes this possible is that portions of the graph are tagged at graph construction time by the function in which they are created using the [`tf.name_scope` decorator](https://www.tensorflow.org/api_docs/python/tf/name_scope), e.g.,
```
def mean_squared_error(x, y):
    with tf.name_scope('mean_squared_error'):
        return tf.reduce_mean((x - y) ** 2)
```
By default, TensorBoard will show a "mean_squared_error" node for each invocation of mean_squared_error rather than the `-`, `**` and `reduce_mean` nodes, but you can click on the node to dynamically expand it if you want to see what happens inside. If `mean_squared_error` were called inside a `evaluate` function, the names just nest even further, e.g., "evaluate/mean_squared_error/subtract".

Dask doesn't have explicit graph objects, which makes this harder. And you still might have an exploding number of tasks when you click to expand, e.g., a `dask.array.top` node. But there might be some way to adjust the data model at the level of dask collections to add some sort of tagging like this, e.g., in the `merge` function. I also think non-static visualizations are a key insight: it's very hard to make this work well with a single fixed graphic.

**Comment by mrocklin** [comment]
@shoyer we often use the names of the keys under the keysplit mapping

    ('add-1234', 0) -> 'add

This convention holds pretty well across the project.

**Comment by shoyer** [comment]
@mrocklin Right, but we lose the higher level structure/groupings, e.g., when `add` is called by two different functions.

**Comment by mrocklin** [comment]
Correct.  I have no plans to address this for now.

**Comment by mrocklin** [comment]
@shoyer can you share some of the visuals from TensorFlow that you're
thinking of or that you have found generally useful?

**Comment by shoyer** [comment]
This page provides a nice overview: https://www.tensorflow.org/get_started/graph_viz

**Comment by jcrist** [comment]
This is something I wrote up and showed @eriknw last week. It's very cheap to compute, but has fail cases. The idea here is that we use the convention of having each key in the graph be a tuple where the first element is a prefix followed by a hash. For dask collections this allows us to group keys from the same operation into a single box, unioning all inputs and outputs. However, this will fail for graphs that don't follow this convention, or graphs that were created with `dask.delayed`, as the tokens for `dask.delayed` graphs are all different.

The code can be found [here](https://gist.github.com/jcrist/dc5b7cedfddff123f2177e5238e566e5).

Example:

```python
In [1]: from vis import simple_vis

In [2]: import numpy as np

In [3]: import dask.array as da

In [4]: x = np.arange(100).reshape((10, 10))

In [5]: dx = da.from_array(x, chunks=(5, 5))

In [6]: res = dx.dot(dx.T).sum(axis=1).mean() + dx.mean(axis=0) * 4

In [7]: simple_vis(res)
Out[7]: <IPython.core.display.Image object>

In [8]: res.visualize()
Out[8]: <IPython.core.display.Image object>
```

The full graph looks like:

![mydask](https://cloud.githubusercontent.com/assets/2783717/23974292/83cb90da-09a8-11e7-83e6-05e819796b71.png)

While the simplified looks like:

![simple](https://cloud.githubusercontent.com/assets/2783717/23974296/86324dd2-09a8-11e7-86a5-caf601a85307.png)

One idea for how to use these would be to reuse the graphviz-to-bokeh code I wrote up 2 years ago to convert these graphs into bokeh plots. We could color the nodes the same as their progressbars, which would allow users to see the progress for each task, as well as where in the general pipeline each task is.

**Comment by mrocklin** [comment]
Using the full hash makes sense to me.  It seems like key conventions give us a few levels to play with.

**Comment by mrocklin** [comment]
A complication here is that the dask.distributed scheduler doesn't get the key as a tuple.  It gets it as a string.  Getting out the group-name is not trivial in the distributed case.

**Comment by jakirkham** [comment]
A more basic solution is to allow this to be user coded. For example users could define small portions of the graph as belonging to some node in the larger graph. IOW this would allow users to nest graphs. I'm not sure if there would be other benefits for Dask ATM, but there may be.

**Comment by jakirkham** [comment]
Probably easier to do today with `HighLevelGraph` objects.

**Comment by mrocklin** [comment]
Right, so HighLevelGraphs objects have a `.visualize` method today, so some of this is done, at least for static graphs.

```python
x.__dask_graph__().visualize()
```

It would be nice to also have diagnostic pages that aggregated output at the key-prefix level (operation names like `add`, `read_csv` and so on).  This isn't quite the same structure as HLGs, but probably provides most of the same benefits.  (reminder, the scheduler doesn't know the high level structure).

**Comment by bicubic** [comment]
Sorry to bump an old issue, but can I ask how the animation in the first post was generated? Is there a code sample anywhere?

**Comment by jcrist** [comment]
I believe I used https://gist.github.com/jcrist/dc5b7cedfddff123f2177e5238e566e5 for this, but that gist is several years old and may no longer work.
